File: sdk/javascript/tests/client.test.ts
import { describe, it, expect, vi, beforeEach } from "vitest";
import { UltraDagClient } from "../src/client.js";
import { UltraDagError, satsToUdag, udagToSats, SATS_PER_UDAG } from "../src/types.js";

// ---------------------------------------------------------------------------
// Mock fetch helper
// ---------------------------------------------------------------------------

function mockFetch(body: unknown, status = 200): typeof globalThis.fetch {
  return vi.fn().mockResolvedValue({
    ok: status >= 200 && status < 300,
    status,
    text: () => Promise.resolve(JSON.stringify(body)),
  }) as unknown as typeof globalThis.fetch;
}

function mockFetchText(text: string, status = 200): typeof globalThis.fetch {
  return vi.fn().mockResolvedValue({
    ok: status >= 200 && status < 300,
    status,
    text: () => Promise.resolve(text),
  }) as unknown as typeof globalThis.fetch;
}

function failingFetch(): typeof globalThis.fetch {
  return vi.fn().mockRejectedValue(new Error("Connection refused")) as unknown as typeof globalThis.fetch;
}

// ---------------------------------------------------------------------------
// Unit helpers
// ---------------------------------------------------------------------------

describe("satsToUdag / udagToSats", () => {
  it("converts sats to UDAG", () => {
    expect(satsToUdag(100_000_000)).toBe(1);
    expect(satsToUdag(50_000_000)).toBe(0.5);
    expect(satsToUdag(0)).toBe(0);
  });

  it("converts UDAG to sats", () => {
    expect(udagToSats(1)).toBe(100_000_000);
    expect(udagToSats(0.5)).toBe(50_000_000);
    expect(udagToSats(0)).toBe(0);
  });

  it("round-trips correctly", () => {
    expect(udagToSats(satsToUdag(123_456_789))).toBe(123_456_789);
  });

  it("SATS_PER_UDAG is 100 million", () => {
    expect(SATS_PER_UDAG).toBe(100_000_000);
  });
});

// ---------------------------------------------------------------------------
// GET endpoints
// ---------------------------------------------------------------------------

describe("UltraDagClient GET endpoints", () => {
  let client: UltraDagClient;
  let fetchFn: ReturnType<typeof vi.fn>;

  function setup(body: unknown, status = 200) {
    const f = mockFetch(body, status);
    fetchFn = f as unknown as ReturnType<typeof vi.fn>;
    client = new UltraDagClient({ baseUrl: "http://node:10333", fetch: f });
  }

  // -- /health ------------------------------------------------------------

  it("getHealth returns status", async () => {
    setup({ status: "ok" });
    const res = await client.getHealth();
    expect(res.status).toBe("ok");
    expect(fetchFn).toHaveBeenCalledWith(
      "http://node:10333/health",
      expect.objectContaining({ method: "GET" }),
    );
  });

  // -- /status ------------------------------------------------------------

  it("getStatus returns node metrics", async () => {
    const data = {
      last_finalized_round: 42,
      peer_count: 3,
      mempool_size: 10,
      total_supply: 2_100_000_000_000_000,
      account_count: 5,
      dag_vertices: 200,
      dag_round: 45,
      dag_tips: 4,
      finalized_count: 180,
      validator_count: 4,
      total_staked: 100_000_000_000,
      active_stakers: 4,
      bootstrap_connected: true,
    };
    setup(data);
    const res = await client.getStatus();
    expect(res.last_finalized_round).toBe(42);
    expect(res.validator_count).toBe(4);
    expect(res.bootstrap_connected).toBe(true);
  });

  // -- /balance/:address --------------------------------------------------

  it("getBalance returns balance info", async () => {
    const data = {
      address: "aa".repeat(32),
      balance: 500_000_000,
      nonce: 3,
      balance_tdag: 5.0,
    };
    setup(data);
    const res = await client.getBalance("aa".repeat(32));
    expect(res.balance).toBe(500_000_000);
    expect(res.nonce).toBe(3);
    expect(fetchFn).toHaveBeenCalledWith(
      `http://node:10333/balance/${"aa".repeat(32)}`,
      expect.anything(),
    );
  });

  // -- /round/:round ------------------------------------------------------

  it("getRound returns vertex list", async () => {
    const vertices = [
      { round: 10, hash: "ab".repeat(32), validator: "cd".repeat(32), reward: 50_000_000_00, tx_count: 2, parent_count: 3 },
    ];
    setup(vertices);
    const res = await client.getRound(10);
    expect(res).toHaveLength(1);
    expect(res[0].round).toBe(10);
    expect(fetchFn).toHaveBeenCalledWith(
      "http://node:10333/round/10",
      expect.anything(),
    );
  });

  // -- /mempool -----------------------------------------------------------

  it("getMempool returns transaction list", async () => {
    const txs = [{ from: "aa".repeat(32), to: "bb".repeat(32), amount: 1000, fee: 100, nonce: 1 }];
    setup(txs);
    const res = await client.getMempool();
    expect(res).toHaveLength(1);
    expect(res[0].amount).toBe(1000);
  });

  // -- /peers -------------------------------------------------------------

  it("getPeers returns peer info", async () => {
    setup({ connected: 3, peers: ["1.2.3.4:9333"], bootstrap_nodes: ["5.6.7.8:9333"] });
    const res = await client.getPeers();
    expect(res.connected).toBe(3);
    expect(res.peers).toContain("1.2.3.4:9333");
  });

  // -- /keygen ------------------------------------------------------------

  it("keygen returns new keypair", async () => {
    setup({ secret_key: "ab".repeat(32), address: "cd".repeat(32) });
    const res = await client.keygen();
    expect(res.secret_key).toHaveLength(64);
    expect(res.address).toHaveLength(64);
  });

  // -- /validators --------------------------------------------------------

  it("getValidators returns validator list", async () => {
    const data = {
      count: 2,
      total_staked: 200_000_000_000,
      validators: [
        { address: "aa".repeat(32), staked: 100_000_000_000, staked_udag: 1000 },
        { address: "bb".repeat(32), staked: 100_000_000_000, staked_udag: 1000 },
      ],
    };
    setup(data);
    const res = await client.getValidators();
    expect(res.count).toBe(2);
    expect(res.validators).toHaveLength(2);
  });

  // -- /stake/:address ----------------------------------------------------

  it("getStakeInfo returns stake details", async () => {
    const data = {
      address: "aa".repeat(32),
      staked: 100_000_000_000,
      staked_udag: 1000,
      unlock_at_round: null,
      is_active_validator: true,
    };
    setup(data);
    const res = await client.getStakeInfo("aa".repeat(32));
    expect(res.is_active_validator).toBe(true);
    expect(res.staked).toBe(100_000_000_000);
  });

  // -- /governance/config -------------------------------------------------

  it("getGovernanceConfig returns config object", async () => {
    setup({ voting_period: 1000, quorum: 50 });
    const res = await client.getGovernanceConfig();
    expect(res).toHaveProperty("voting_period");
  });

  // -- /proposals ---------------------------------------------------------

  it("getProposals returns proposals list", async () => {
    setup({ count: 1, proposals: [{ id: 1, title: "Test" }] });
    const res = await client.getProposals();
    expect(res.count).toBe(1);
    expect(res.proposals).toHaveLength(1);
  });

  // -- /proposal/:id ------------------------------------------------------

  it("getProposal returns proposal details", async () => {
    setup({ id: 5, title: "Change round time" });
    const res = await client.getProposal(5);
    expect(res.id).toBe(5);
    expect(fetchFn).toHaveBeenCalledWith(
      "http://node:10333/proposal/5",
      expect.anything(),
    );
  });

  // -- /vote/:proposal_id/:address ----------------------------------------

  it("getVote returns vote info", async () => {
    setup({ vote: "yes", weight: 100 });
    const res = await client.getVote(5, "aa".repeat(32));
    expect(res).toHaveProperty("vote");
    expect(fetchFn).toHaveBeenCalledWith(
      `http://node:10333/vote/5/${"aa".repeat(32)}`,
      expect.anything(),
    );
  });
});

// ---------------------------------------------------------------------------
// POST endpoints
// ---------------------------------------------------------------------------

describe("UltraDagClient POST endpoints", () => {
  let client: UltraDagClient;
  let fetchFn: ReturnType<typeof vi.fn>;

  function setup(body: unknown, status = 200) {
    const f = mockFetch(body, status);
    fetchFn = f as unknown as ReturnType<typeof vi.fn>;
    client = new UltraDagClient({ baseUrl: "http://node:10333", fetch: f });
  }

  // -- /tx ----------------------------------------------------------------

  it("sendTransaction posts tx and returns result", async () => {
    const response = {
      hash: "ff".repeat(32),
      from: "aa".repeat(32),
      to: "bb".repeat(32),
      amount: 1_000_000,
      fee: 1000,
      nonce: 1,
    };
    setup(response);
    const res = await client.sendTransaction({
      secret_key: "cc".repeat(32),
      to: "bb".repeat(32),
      amount: 1_000_000,
      fee: 1000,
    });
    expect(res.hash).toBe("ff".repeat(32));
    expect(res.amount).toBe(1_000_000);
    // Verify POST body
    const callArgs = fetchFn.mock.calls[0];
    expect(callArgs[1].method).toBe("POST");
    const sentBody = JSON.parse(callArgs[1].body);
    expect(sentBody.to).toBe("bb".repeat(32));
    expect(sentBody.amount).toBe(1_000_000);
  });

  // -- /faucet ------------------------------------------------------------

  it("faucet sends address and amount", async () => {
    const response = {
      tx_hash: "dd".repeat(32),
      from: "fa".repeat(32),
      to: "aa".repeat(32),
      amount: 100_000_000,
      amount_udag: 1.0,
      nonce: 0,
    };
    setup(response);
    const res = await client.faucet("aa".repeat(32), 100_000_000);
    expect(res.tx_hash).toBe("dd".repeat(32));
    expect(res.amount_udag).toBe(1.0);
    const sentBody = JSON.parse(fetchFn.mock.calls[0][1].body);
    expect(sentBody.address).toBe("aa".repeat(32));
    expect(sentBody.amount).toBe(100_000_000);
  });

  // -- /stake -------------------------------------------------------------

  it("stake posts staking request", async () => {
    const response = {
      status: "staked",
      tx_hash: "ee".repeat(32),
      address: "aa".repeat(32),
      amount: 1_000_000_000_000,
      amount_udag: 10000,
      nonce: 2,
      note: "Staked 10000 UDAG",
    };
    setup(response);
    const res = await client.stake({
      secret_key: "cc".repeat(32),
      amount: 1_000_000_000_000,
    });
    expect(res.status).toBe("staked");
    expect(res.amount_udag).toBe(10000);
  });

  // -- /unstake -----------------------------------------------------------

  it("unstake posts unstake request", async () => {
    const response = {
      status: "unstaking",
      tx_hash: "ff".repeat(32),
      address: "aa".repeat(32),
      unlock_at_round: 5000,
      nonce: 3,
      note: "Unstaking, cooldown 2016 rounds",
    };
    setup(response);
    const res = await client.unstake({ secret_key: "cc".repeat(32) });
    expect(res.status).toBe("unstaking");
    expect(res.unlock_at_round).toBe(5000);
  });

  // -- /proposal ----------------------------------------------------------

  it("createProposal posts proposal", async () => {
    setup({ id: 1, status: "created" });
    const res = await client.createProposal({
      proposer_secret: "cc".repeat(32),
      title: "Change round time",
      description: "Reduce to 3 seconds",
      proposal_type: "parameter_change",
      parameter_name: "round_ms",
      parameter_value: "3000",
    });
    expect(res).toHaveProperty("id");
    const sentBody = JSON.parse(fetchFn.mock.calls[0][1].body);
    expect(sentBody.title).toBe("Change round time");
    expect(sentBody.parameter_name).toBe("round_ms");
  });

  // -- /vote --------------------------------------------------------------

  it("castVote posts vote", async () => {
    setup({ status: "voted" });
    const res = await client.castVote({
      voter_secret: "cc".repeat(32),
      proposal_id: 1,
      vote: "yes",
    });
    expect(res).toHaveProperty("status");
    const sentBody = JSON.parse(fetchFn.mock.calls[0][1].body);
    expect(sentBody.proposal_id).toBe(1);
    expect(sentBody.vote).toBe("yes");
  });
});

// ---------------------------------------------------------------------------
// Error handling
// ---------------------------------------------------------------------------

describe("UltraDagClient error handling", () => {
  it("throws UltraDagError on HTTP error status", async () => {
    const f = mockFetch({ error: "Not found" }, 404);
    const client = new UltraDagClient({ baseUrl: "http://node:10333", fetch: f });
    await expect(client.getBalance("aa".repeat(32))).rejects.toThrow(UltraDagError);
    try {
      await client.getBalance("aa".repeat(32));
    } catch (e) {
      expect(e).toBeInstanceOf(UltraDagError);
      const err = e as UltraDagError;
      expect(err.status).toBe(404);
      expect(err.body).toContain("Not found");
    }
  });

  it("throws UltraDagError on network failure", async () => {
    const f = failingFetch();
    const client = new UltraDagClient({ baseUrl: "http://node:10333", fetch: f });
    await expect(client.getHealth()).rejects.toThrow(UltraDagError);
    try {
      await client.getHealth();
    } catch (e) {
      expect(e).toBeInstanceOf(UltraDagError);
      expect((e as UltraDagError).message).toContain("Connection refused");
    }
  });

  it("throws UltraDagError on invalid JSON response", async () => {
    const f = mockFetchText("not json", 200);
    const client = new UltraDagClient({ baseUrl: "http://node:10333", fetch: f });
    await expect(client.getHealth()).rejects.toThrow(UltraDagError);
    try {
      await client.getHealth();
    } catch (e) {
      expect(e).toBeInstanceOf(UltraDagError);
      expect((e as UltraDagError).message).toContain("Invalid JSON");
    }
  });

  it("throws UltraDagError on server 500 error", async () => {
    const f = mockFetchText("Internal Server Error", 500);
    const client = new UltraDagClient({ baseUrl: "http://node:10333", fetch: f });
    await expect(client.getStatus()).rejects.toThrow(UltraDagError);
    try {
      await client.getStatus();
    } catch (e) {
      const err = e as UltraDagError;
      expect(err.status).toBe(500);
    }
  });
});

// ---------------------------------------------------------------------------
// Client configuration
// ---------------------------------------------------------------------------

describe("UltraDagClient configuration", () => {
  it("uses default base URL when none provided", async () => {
    const f = mockFetch({ status: "ok" });
    const fetchFn = f as unknown as ReturnType<typeof vi.fn>;
    const client = new UltraDagClient({ fetch: f });
    await client.getHealth();
    expect(fetchFn.mock.calls[0][0]).toBe("http://localhost:10333/health");
  });

  it("strips trailing slash from base URL", async () => {
    const f = mockFetch({ status: "ok" });
    const fetchFn = f as unknown as ReturnType<typeof vi.fn>;
    const client = new UltraDagClient({ baseUrl: "http://node:10333/", fetch: f });
    await client.getHealth();
    expect(fetchFn.mock.calls[0][0]).toBe("http://node:10333/health");
  });
});

File: sdk/rust/tests/client_test.rs
use mockito::{Matcher, Server};
use ultradag_sdk::UltraDagClient;

fn make_client(server: &Server) -> UltraDagClient {
    UltraDagClient::new(&server.url())
}

// ---------------------------------------------------------------------------
// GET endpoints
// ---------------------------------------------------------------------------

#[test]
fn test_health() {
    let mut server = Server::new();
    let mock = server
        .mock("GET", "/health")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(r#"{"status":"ok"}"#)
        .create();

    let client = make_client(&server);
    let resp = client.health().unwrap();
    assert_eq!(resp.status, "ok");
    mock.assert();
}

#[test]
fn test_status() {
    let mut server = Server::new();
    let body = serde_json::json!({
        "last_finalized_round": 42,
        "peer_count": 3,
        "mempool_size": 10,
        "total_supply": 2100000000000000u64,
        "account_count": 5,
        "dag_vertices": 200,
        "dag_round": 50,
        "dag_tips": 4,
        "finalized_count": 180,
        "validator_count": 4,
        "total_staked": 1000000000000u64,
        "active_stakers": 4,
        "bootstrap_connected": true
    });
    let mock = server
        .mock("GET", "/status")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(body.to_string())
        .create();

    let client = make_client(&server);
    let resp = client.status().unwrap();
    assert_eq!(resp.last_finalized_round, Some(42));
    assert_eq!(resp.peer_count, 3);
    assert_eq!(resp.dag_round, 50);
    assert!(resp.bootstrap_connected);
    mock.assert();
}

#[test]
fn test_balance() {
    let mut server = Server::new();
    let addr = "a".repeat(64);
    let body = serde_json::json!({
        "address": addr,
        "balance": 500000000u64,
        "nonce": 3,
        "balance_tdag": 5.0
    });
    let mock = server
        .mock("GET", format!("/balance/{addr}").as_str())
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(body.to_string())
        .create();

    let client = make_client(&server);
    let resp = client.balance(&addr).unwrap();
    assert_eq!(resp.balance, 500000000);
    assert_eq!(resp.nonce, 3);
    mock.assert();
}

#[test]
fn test_balance_invalid_address() {
    let server = Server::new();
    let client = make_client(&server);
    let err = client.balance("tooshort").unwrap_err();
    assert!(matches!(err, ultradag_sdk::UltraDagError::InvalidAddress(_)));
}

#[test]
fn test_round() {
    let mut server = Server::new();
    let body = serde_json::json!([
        {
            "round": 10,
            "hash": "abcd1234",
            "validator": "val1",
            "reward": 5000000000u64,
            "tx_count": 2,
            "parent_count": 3
        }
    ]);
    let mock = server
        .mock("GET", "/round/10")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(body.to_string())
        .create();

    let client = make_client(&server);
    let resp = client.round(10).unwrap();
    assert_eq!(resp.len(), 1);
    assert_eq!(resp[0].round, 10);
    assert_eq!(resp[0].tx_count, 2);
    mock.assert();
}

#[test]
fn test_mempool() {
    let mut server = Server::new();
    let mock = server
        .mock("GET", "/mempool")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body("[]")
        .create();

    let client = make_client(&server);
    let resp = client.mempool().unwrap();
    assert!(resp.is_empty());
    mock.assert();
}

#[test]
fn test_keygen() {
    let mut server = Server::new();
    let body = serde_json::json!({
        "secret_key": "ab".repeat(32),
        "address": "cd".repeat(32)
    });
    let mock = server
        .mock("GET", "/keygen")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(body.to_string())
        .create();

    let client = make_client(&server);
    let resp = client.keygen().unwrap();
    assert_eq!(resp.secret_key.len(), 64);
    assert_eq!(resp.address.len(), 64);
    mock.assert();
}

#[test]
fn test_peers() {
    let mut server = Server::new();
    let body = serde_json::json!({
        "connected": 2,
        "peers": ["1.2.3.4:9333", "5.6.7.8:9333"],
        "bootstrap_nodes": [
            {"address": "1.2.3.4:9333", "connected": true}
        ]
    });
    let mock = server
        .mock("GET", "/peers")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(body.to_string())
        .create();

    let client = make_client(&server);
    let resp = client.peers().unwrap();
    assert_eq!(resp.connected, 2);
    assert_eq!(resp.peers.len(), 2);
    assert_eq!(resp.bootstrap_nodes.len(), 1);
    assert!(resp.bootstrap_nodes[0].connected);
    mock.assert();
}

#[test]
fn test_validators() {
    let mut server = Server::new();
    let body = serde_json::json!({
        "count": 1,
        "total_staked": 1000000000000u64,
        "validators": [
            {"address": "a".repeat(64), "stake": 1000000000000u64}
        ]
    });
    let mock = server
        .mock("GET", "/validators")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(body.to_string())
        .create();

    let client = make_client(&server);
    let resp = client.validators().unwrap();
    assert_eq!(resp.count, 1);
    assert_eq!(resp.validators[0].stake, 1000000000000);
    mock.assert();
}

#[test]
fn test_stake_info() {
    let mut server = Server::new();
    let addr = "b".repeat(64);
    let body = serde_json::json!({
        "address": addr,
        "staked": 500000000000u64,
        "staked_udag": 5000.0,
        "unlock_at_round": null,
        "is_active_validator": true
    });
    let mock = server
        .mock("GET", format!("/stake/{addr}").as_str())
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(body.to_string())
        .create();

    let client = make_client(&server);
    let resp = client.stake_info(&addr).unwrap();
    assert_eq!(resp.staked, 500000000000);
    assert!(resp.is_active_validator);
    assert!(resp.unlock_at_round.is_none());
    mock.assert();
}

#[test]
fn test_governance_config() {
    let mut server = Server::new();
    let body = serde_json::json!({"min_stake": 10000});
    let mock = server
        .mock("GET", "/governance/config")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(body.to_string())
        .create();

    let client = make_client(&server);
    let resp = client.governance_config().unwrap();
    assert_eq!(resp["min_stake"], 10000);
    mock.assert();
}

#[test]
fn test_proposals() {
    let mut server = Server::new();
    let body = serde_json::json!({"count": 0, "proposals": []});
    let mock = server
        .mock("GET", "/proposals")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(body.to_string())
        .create();

    let client = make_client(&server);
    let resp = client.proposals().unwrap();
    assert_eq!(resp.count, 0);
    mock.assert();
}

#[test]
fn test_proposal() {
    let mut server = Server::new();
    let body = serde_json::json!({"id": 1, "title": "Test"});
    let mock = server
        .mock("GET", "/proposal/1")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(body.to_string())
        .create();

    let client = make_client(&server);
    let resp = client.proposal(1).unwrap();
    assert_eq!(resp["id"], 1);
    mock.assert();
}

#[test]
fn test_vote_info() {
    let mut server = Server::new();
    let addr = "c".repeat(64);
    let body = serde_json::json!({"voted": true});
    let mock = server
        .mock("GET", format!("/vote/1/{addr}").as_str())
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(body.to_string())
        .create();

    let client = make_client(&server);
    let resp = client.vote_info(1, &addr).unwrap();
    assert_eq!(resp["voted"], true);
    mock.assert();
}

// ---------------------------------------------------------------------------
// POST endpoints
// ---------------------------------------------------------------------------

#[test]
fn test_send_tx() {
    let mut server = Server::new();
    let body = serde_json::json!({
        "hash": "deadbeef",
        "from": "f".repeat(64),
        "to": "e".repeat(64),
        "amount": 100000000u64,
        "fee": 100000u64,
        "nonce": 1
    });
    let mock = server
        .mock("POST", "/tx")
        .match_header("content-type", "application/json")
        .match_body(Matcher::Any)
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(body.to_string())
        .create();

    let client = make_client(&server);
    let resp = client
        .send_tx(&"ab".repeat(16), &"e".repeat(64), 100000000, 100000)
        .unwrap();
    assert_eq!(resp.amount, 100000000);
    assert_eq!(resp.nonce, 1);
    mock.assert();
}

#[test]
fn test_faucet() {
    let mut server = Server::new();
    let body = serde_json::json!({"success": true});
    let mock = server
        .mock("POST", "/faucet")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(body.to_string())
        .create();

    let client = make_client(&server);
    let resp = client.faucet(&"d".repeat(64), 1000000000).unwrap();
    assert_eq!(resp["success"], true);
    mock.assert();
}

#[test]
fn test_stake() {
    let mut server = Server::new();
    let body = serde_json::json!({"ok": true});
    let mock = server
        .mock("POST", "/stake")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(body.to_string())
        .create();

    let client = make_client(&server);
    let resp = client.stake(&"ab".repeat(16), 1000000000000).unwrap();
    assert_eq!(resp["ok"], true);
    mock.assert();
}

#[test]
fn test_unstake() {
    let mut server = Server::new();
    let body = serde_json::json!({"ok": true});
    let mock = server
        .mock("POST", "/unstake")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(body.to_string())
        .create();

    let client = make_client(&server);
    let resp = client.unstake(&"ab".repeat(16)).unwrap();
    assert_eq!(resp["ok"], true);
    mock.assert();
}

#[test]
fn test_submit_proposal() {
    let mut server = Server::new();
    let body = serde_json::json!({"id": 1});
    let mock = server
        .mock("POST", "/proposal")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(body.to_string())
        .create();

    let client = make_client(&server);
    let req = ultradag_sdk::ProposalRequest {
        proposer_secret: "ab".repeat(16),
        title: "Test".into(),
        description: "A test proposal".into(),
        proposal_type: "parameter_change".into(),
        parameter_name: Some("round_ms".into()),
        parameter_value: Some("3000".into()),
        fee: None,
    };
    let resp = client.submit_proposal(&req).unwrap();
    assert_eq!(resp["id"], 1);
    mock.assert();
}

#[test]
fn test_vote() {
    let mut server = Server::new();
    let body = serde_json::json!({"ok": true});
    let mock = server
        .mock("POST", "/vote")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(body.to_string())
        .create();

    let client = make_client(&server);
    let req = ultradag_sdk::VoteRequest {
        voter_secret: "ab".repeat(16),
        proposal_id: 1,
        vote: true,
        fee: None,
    };
    let resp = client.vote(&req).unwrap();
    assert_eq!(resp["ok"], true);
    mock.assert();
}

// ---------------------------------------------------------------------------
// Error handling
// ---------------------------------------------------------------------------

#[test]
fn test_api_error_404() {
    let mut server = Server::new();
    let mock = server
        .mock("GET", "/balance/aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa")
        .with_status(404)
        .with_body("not found")
        .create();

    let client = make_client(&server);
    let addr = "a".repeat(64);
    let err = client.balance(&addr).unwrap_err();
    match err {
        ultradag_sdk::UltraDagError::Api { status, message } => {
            assert_eq!(status, 404);
            assert_eq!(message, "not found");
        }
        other => panic!("expected Api error, got: {other:?}"),
    }
    mock.assert();
}

#[test]
fn test_json_parse_error() {
    let mut server = Server::new();
    let mock = server
        .mock("GET", "/status")
        .with_status(200)
        .with_body("not json at all")
        .create();

    let client = make_client(&server);
    let err = client.status().unwrap_err();
    assert!(matches!(err, ultradag_sdk::UltraDagError::Json(_)));
    mock.assert();
}

#[test]
fn test_connection_error() {
    // No server running on this port
    let client = UltraDagClient::new("http://127.0.0.1:19999");
    let err = client.health().unwrap_err();
    assert!(matches!(err, ultradag_sdk::UltraDagError::Http(_)));
}
